src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -175,7 +180,33 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
                 // for empty array.
                 return false;
             }
-            _ => {
+            ty::Bool
+            | ty::Char
+            | ty::Int(_)
+            | ty::Uint(_)
+            | ty::Float(_)
+            | ty::Str
+            | ty::Never
+            | ty::Error => {
+                // These primitive types are always structural match.
+                //
+                // `Never` is kind of special here, but as it is not inhabitable, this should be fine.
+                return false;
+            }
+
+            ty::Array(..)
+            | ty::Slice(_)
+            | ty::Ref(..)
+            | ty::Closure(..)
+            | ty::Generator(..)
+            | ty::GeneratorWitness(..)
+            | ty::Tuple(..)
+            | ty::Projection(..)
+            | ty::UnnormalizedProjection(..)
+            | ty::Opaque(..)
+            | ty::Bound(..)
+            | ty::Placeholder(_)
+            | ty::Infer(_) => {
                 ty.super_visit_with(self);
```

Comment:
So these should be "FIXME: figure out what to do", if I understand the PR description correctly?

Reply:
jup

Reply:
They now are let's either emit a bug or an error, which is a lot cleaner :laughing: